### PR TITLE
[Diagnostics & Tests] SR-6052 Prevent nil capitalization

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Apart from preventing capitalization in Xcode, this provides better semantic background by surrounding 'nil' in ticks when it is referred to as a literal.
Also added missing tests for certain cases involving nil capitalization.

Resolves [SR-6052](https://bugs.swift.org/browse/SR-6052).

### CODE REVIEW

**Comment by CodaFi** [comment]
To keep the diff cleaner, could you split the formatting commit into a different PR?

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -327,13 +327,13 @@ ERROR(cannot_throw_error_code,none,
       "instance", (Type, Type))
 
 ERROR(cannot_throw_nil,none,
-      "cannot infer concrete Error for thrown 'nil' value", ())
```
Comment (hamishknight) [inline include/swift/AST/DiagnosticsSema.def]: Did you mean to remove only the closing `'` here?

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -327,13 +327,13 @@ ERROR(cannot_throw_error_code,none,
...
+      "cannot infer concrete Error for thrown 'nil value", ())
```
Comment (hamishknight) [inline include/swift/AST/DiagnosticsSema.def]: Did you mean to remove only the closing `'` here?

Comment (AnthonyLatsis) [reply include/swift/AST/DiagnosticsSema.def]: Thanks.

**Comment by AnthonyLatsis** [comment]
Sure, reset.

**Review by xedin — COMMENTED** [review]
I think @jrose-apple is a better person to review this but to me it seems like ticking nil doesn't make sense for some diagnostics e.g. - `comparing non-optional value of type 'Any.Type' to 'nil' always returns false`

**Comment by AnthonyLatsis** [comment]
Yes, at first I wanted to update only those that lead to capitalization. Seems I ended up touching a couple of controversial cases along the way. I agree and can exclude them any time. I think Jordan will agree as well.

**Review by jrose-apple — APPROVED** [review]
Okay, yes, arguably all the ones you've added quotes to could refer to the literal expression rather than the value, so let's just go with this. (I appreciate the removal from "non-nil" as well.)

**Comment by AnthonyLatsis** [comment]
@jrose-apple Looks like it hasn't started building

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
Excuse the double commit, I am running build-script.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@AnthonyLatsis Don't worry, I'll squash before merging.

**Comment by AnthonyLatsis** [comment]
@xedin I take it this requires write access since the commits are squashed when rebased onto master?

**Comment by jrose-apple** [comment]
Got it for you. Thanks, Anthony!

**Pull request merged**